### PR TITLE
[SPARK-34547][SQL] Only use metadata columns for resolution as last resort

### PR DESCRIPTION
**Pull request by karenfeng**
### What changes were proposed in this pull request?

Today, child expressions may be resolved based on "real" or metadata output attributes. We should prefer the real attribute during resolution if one exists.

### Why are the changes needed?

Today, attempting to resolve an expression when there is a "real" output attribute and a metadata attribute with the same name results in resolution failure. This is likely unexpected, as the user may not know about the metadata attribute.

### Does this PR introduce _any_ user-facing change?

Yes. Previously, the user would see an error message when resolving a column with the same name as a "real" output attribute and a metadata attribute as below:
```
org.apache.spark.sql.AnalysisException: Reference 'index' is ambiguous, could be: testcat.ns1.ns2.tableTwo.index, testcat.ns1.ns2.tableOne.index.; line 1 pos 71
at org.apache.spark.sql.catalyst.expressions.package$AttributeSeq.resolve(package.scala:363)
at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveChildren(LogicalPlan.scala:107)
```

Now, resolution succeeds and provides the "real" output attribute.

### How was this patch tested?

Added a unit test.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 1]
```diff
@@ -2492,6 +2492,35 @@ class DataSourceV2SQLSuite
     }
   }
 
+  test("SPARK-34547: metadata columns are resolved last") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2495]: I'm a bit confused. We have a test `SPARK-31255: Projects data column when metadata column has the same name` in this file, which tests when column name and metadata column name have conflicts. Why does that test work?

Comment (karenfeng) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2495]: This is covered at the [level](https://github.com/apache/spark/blob/62737e140c7b04805726a33c392c297335db7b45/sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala#L59) of the DatasourceV2Relation, but does not apply above it.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2495]: Ah I see! Can we remove the hack in `DatasourceV2Relation` now?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 2]
```diff
@@ -2492,6 +2492,35 @@ class DataSourceV2SQLSuite
...
+    val t1 = s"${catalogAndNamespace}tableOne"
+    val t2 = s"${catalogAndNamespace}tableTwo"
+    withTable(t1, t2) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string) USING $v2Format " +
+        "PARTITIONED BY (bucket(4, id), id)")
+      sql(s"INSERT INTO $t1 VALUES (1, 'a'), (2, 'b'), (3, 'c')")
+      sql(s"CREATE TABLE $t2 (id bigint, index int) USING $v2Format")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2502]: I think the test is to join a v2 table with a relation, while the relation output has name conflicts with the v2 table's metadata output. It doesn't matter what the relation is and we can use a temp view to simplify the test.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 1 above)
Comment (karenfeng) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2495]: We can remove it from `metadataOutput`, but I don't believe we can remove it in `withMetadataColumns` due to the output containing a name collision.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Pull request closed without merging**